File: packages/node/src/plugins/i18n.ts
/**
 * Built-in internationalization (i18n) plugin.
 *
 * Demonstrates the plugin extension surface: a factory returning a {@link Plugin} that wires a
 * middleware to attach per-update translation sessions, with per-user locale preference stored
 * in `context.user_data` (and therefore persisted automatically).
 *
 * @packageDocumentation
 */

import type { Plugin } from "../kernel/plugin.js";
import type { Application } from "../kernel/app.js";
import type { CallbackContext } from "../kernel/context.js";

/**
 * Configuration for the {@link i18n} plugin.
 *
 * @example
 * ```ts
 * const options: I18nOptions = {
 *   default_locale: "en",
 *   locales: {
 *     en: { hello: "Hello, {name}!" },
 *     vi: { hello: "Xin chào, {name}!" },
 *   },
 * };
 * ```
 */
export interface I18nOptions {
  /**
   * Locale used when the user's preferred locale has no translations.
   */
  default_locale: string;

  /**
   * Translation tables keyed by locale code, then by message key.
   */
  locales: Record<string, Record<string, string>>;

  /**
   * Key inside `context.user_data` where the chosen locale is stored.
   *
   * @defaultValue `"_telebot_locale"`
   */
  locale_key?: string;
}

/**
 * Per-update translation session attached to the {@link CallbackContext} by the i18n middleware.
 */
export interface I18nSession {
  /**
   * Locale code resolved for the current update.
   */
  readonly locale: string;

  /**
   * Translates a message key for the resolved locale, falling back to the default locale, then
   * to the key itself. `{placeholder}` tokens are replaced from `params`.
   *
   * @param key - Message key in the translation tables.
   * @param params - Optional placeholder values.
   * @returns The translated string.
   */
  t(key: string, params?: Record<string, string | number>): string;

  /**
   * Persists the user's locale preference for all future updates.
   *
   * @param locale - Locale code to use from now on.
   */
  setLocale(locale: string): void;
}

const sessions = new WeakMap<CallbackContext, I18nSession>();

/**
 * Retrieves the i18n session attached to a context by the {@link i18n} plugin.
 *
 * @param context - The callback context of the current update.
 * @returns The session, or `undefined` if the plugin is not installed.
 *
 * @example
 * ```ts
 * app.on("message", async (update, context) => {
 *   const t = i18nFor(context)?.t("hello", { name: "Nam" });
 *   await context.reply(t ?? "no i18n");
 * });
 * ```
 */
export function i18nFor(context: CallbackContext): I18nSession | undefined {
  return sessions.get(context);
}

/**
 * Creates the built-in internationalization {@link Plugin}.
 *
 * The middleware resolves a locale per update (stored preference, then Telegram
 * `language_code`, then `default_locale`) and attaches an {@link I18nSession} readable through
 * {@link i18nFor}.
 *
 * @param options - Translation tables and locale configuration.
 * @returns A {@link Plugin} to install with `app.usePlugin(...)`.
 *
 * @example
 * ```ts
 * import { plugins, i18nFor } from "telebot-ts";
 *
 * app.usePlugin(
 *   plugins.i18n({
 *     default_locale: "en",
 *     locales: { en: { hello: "Hello!" }, vi: { hello: "Xin chào!" } },
 *   }),
 * );
 * app.on("message", async (update, context) => {
 *   await context.reply(i18nFor(context)!.t("hello"));
 * });
 * ```
 */
export function i18n(options: I18nOptions): Plugin {
  const localeKey = options.locale_key ?? "_telebot_locale";
  const fallbackTable = options.locales[options.default_locale] ?? {};

  return {
    name: "telebot-plugin-i18n",
    install(app: Application) {
      app.use(async (context, next) => {
        const stored = context.user_data?.[localeKey];
        const preferred =
          typeof stored === "string"
            ? stored
            : (context.update?.effective_user?.language_code ?? options.default_locale);
        const locale =
          options.locales[preferred] !== undefined ? preferred : options.default_locale;
        const table = options.locales[locale] ?? fallbackTable;

        sessions.set(context, {
          locale,
          t(key, params) {
            let text = table[key] ?? fallbackTable[key] ?? key;
            if (params) {
              for (const [name, value] of Object.entries(params)) {
                text = text.split(`{${name}}`).join(String(value));
              }
            }
            return text;
          },
          setLocale(nextLocale) {
            const data = context.user_data;
            if (data) {
              data[localeKey] = nextLocale;
            }
          },
        });

        await next();
      });
    },
  };
}

File: packages/python/src/telebot_py/plugins/i18n.py
"""Built-in internationalization plugin (T045).

Parity with ``packages/node/src/plugins/i18n.ts``: per-update locale
resolution (stored preference in ``context.user_data``, then the sender's
Telegram ``language_code``, then the default locale), missing-key fallback
to the default locale table and finally the key itself, and
``{placeholder}`` interpolation. The per-user preference lives in
``user_data`` and therefore rides on persistence automatically.
"""

from __future__ import annotations

import typing as t
from collections.abc import Mapping, Sequence

from telebot_py.plugins.plugin import Plugin

if t.TYPE_CHECKING:
    from telebot_py.kernel.context import CallbackContext

DEFAULT_LOCALE_KEY = "_telebot_locale"


class I18nSession:
    """Per-update translation session for one resolved locale.

    Created by :meth:`I18nPlugin.session_for`; not constructed directly.

    Attributes:
        locale: The locale code resolved for the current update.
    """

    def __init__(self, plugin: I18nPlugin, context: CallbackContext, locale: str) -> None:
        """Bind the session to its plugin, context, and resolved locale.

        Args:
            plugin: The owning i18n plugin (supplies the locale tables).
            context: The update context; ``user_data`` stores the preference.
            locale: The resolved locale code.
        """
        self.locale = locale
        self._plugin = plugin
        self._context = context

    def t(self, key: str, params: Mapping[str, str | int] | None = None) -> str:
        """Translate a message key for this session's locale.

        Falls back to the default locale's table, then to the key itself.
        ``{placeholder}`` tokens are replaced from ``params``.

        Args:
            key: Message key in the translation tables.
            params: Optional placeholder values.

        Returns:
            The translated, interpolated string.

        Example:
            >>> session.t("hello", {"name": "Nam"})
            'Xin chào, Nam!'
        """
        return self._plugin.translate(self.locale, key, params)

    def set_locale(self, locale: str) -> None:
        """Persist the user's locale preference for all future updates.

        Writes into ``context.user_data``; when the update carries no user
        data this is a harmless no-op (node parity).

        Args:
            locale: Locale code to use from now on.
        """
        data = self._context.user_data
        if data is not None:
            data[self._plugin.locale_key] = locale


class I18nPlugin(Plugin):
    """Internationalization plugin: locale tables plus per-user preference.

    The kernel integration pass can attach a session per update via
    :meth:`session_for`; bot handlers translate through the session's
    :meth:`I18nSession.t`.

    Example:
        >>> plugin = I18nPlugin(
        ...     default_locale="en",
        ...     locales={"en": {"hello": "Hello, {name}!"}},
        ... )
        >>> session = plugin.session_for(context)
        >>> await context.bot.send_message(chat_id, session.t("hello", {"name": "Nam"}))

    Attributes:
        default_locale: Locale used when the user's preferred locale has no
            translation table.
        locales: Translation tables keyed by locale code, then message key.
        locale_key: Key inside ``context.user_data`` storing the preference.
    """

    depends_on: Sequence[str] = ()

    def __init__(
        self,
        *,
        default_locale: str,
        locales: Mapping[str, Mapping[str, str]],
        locale_key: str = DEFAULT_LOCALE_KEY,
        name: str = "telebot-plugin-i18n",
    ) -> None:
        """Configure the translation tables.

        Args:
            default_locale: Fallback locale for missing tables/keys.
            locales: Translation tables keyed by locale code.
            locale_key: ``user_data`` key storing the chosen locale.
            name: Plugin identifier.
        """
        self.name = name
        self.default_locale = default_locale
        self.locales: Mapping[str, Mapping[str, str]] = locales
        self.locale_key = locale_key

    def session_for(self, context: CallbackContext) -> I18nSession:
        """Resolve the locale for an update and open a translation session.

        Resolution order (node parity): stored preference in ``user_data``,
        then the sender's ``language_code``, then ``default_locale``; a
        preferred locale without a translation table falls back to the
        default.

        Args:
            context: The callback context of the current update.

        Returns:
            A session bound to the resolved locale.
        """
        stored: object = None
        if context.user_data is not None:
            stored = context.user_data.get(self.locale_key)
        preferred: str | None
        if isinstance(stored, str) and stored:
            preferred = stored
        else:
            user = context.update.effective_user
            preferred = user.language_code if user is not None and user.language_code else None
        locale = preferred if preferred is not None and preferred in self.locales else None
        return I18nSession(self, context, locale if locale is not None else self.default_locale)

    def translate(
        self,
        locale: str,
        key: str,
        params: Mapping[str, str | int] | None = None,
    ) -> str:
        """Translate a key for an explicit locale.

        Args:
            locale: Locale code to look up first.
            key: Message key in the translation tables.
            params: Optional placeholder values.

        Returns:
            The translated string; falls back to the default locale's table,
            then to the key itself, with placeholders interpolated.
        """
        text = self.locales.get(locale, {}).get(key)
        if text is None:
            text = self.locales.get(self.default_locale, {}).get(key)
        if text is None:
            text = key
        if params:
            for param, value in params.items():
                text = text.replace(f"{{{param}}}", str(value))
        return text
